### PR TITLE
refactor the user location logic on the Events page

### PR DESCRIPTION
**Pull request by s-cheng**
Based on https://github.com/ForumMagnum/ForumMagnum/pull/4555, where I created the Events page.

I think the logic around the user's location works in the original PR, but it was pretty complicated due to trying to integrate with some previously existing code. In this PR, I've refactored much of that logic.

The main changes are:
1. I added checking browser local storage to the `useUserLocation()` hook.
2. This means that, on the Events page, there's now only **one** location-related variable.
3. This also means that we can use the location in local storage on the rest of the forum - I added it to the Community page and the sidebar events list.
4. I also fixed a small issue where a logged out user coming to the Events page for the first time saw a flash of the default events list before the browser gave us their location.

### CODE REVIEW

File: packages/lesswrong/lib/collections/users/helpers.ts [hunk 1]
```diff
@@ -306,47 +307,84 @@ export const userGetLocation = (currentUser: UsersCurrent|DbUser|null): {
...
+ *
+ * @returns {Object} location
+ * @returns {number} location.lat - The user's latitude.
+ * @returns {number} location.lng - The user's longitude.
+ * @returns {boolean} location.loading - Indicates that we might have a known location later.
+ * @returns {boolean} location.known - If false, then we're returning the default location instead of the user's location.
+ * @returns {string} location.label - The string description of the location (ex: Cambridge, MA, USA).
+ * @returns {Function} location.setLocation - Function to set the location directly.
+ */
+export const useUserLocation = (currentUser: UsersCurrent|DbUser|null, dontAsk?: boolean): {
```
Comment (s-cheng) [inline packages/lesswrong/lib/collections/users/helpers.ts:334]: This is quite large now, maybe I should move it somewhere else? Not sure where

Comment (jpaddison3) [reply packages/lesswrong/lib/collections/users/helpers.ts:334]: I'm mildly ok with this.

File: packages/lesswrong/lib/collections/users/helpers.ts [hunk 2]
```diff
@@ -306,47 +307,84 @@ export const userGetLocation = (currentUser: UsersCurrent|DbUser|null): {
...
   
   const currentUserLat = currentUser && currentUser.mongoLocation && currentUser.mongoLocation.coordinates[1]
   const currentUserLng = currentUser && currentUser.mongoLocation && currentUser.mongoLocation.coordinates[0]
 
   const [location, setLocation] = useState(() => {
-    if (currentUserLat && currentUserLng) {
+    if (currentUserLat && currentUserLng && currentUser?.location) {
       // First return a location from the user profile, if set
-      return {lat: currentUserLat, lng: currentUserLng, loading: false, known: true}
+      return {lat: currentUserLat, lng: currentUserLng, loading: false, known: true, label: currentUser.location}
```
Comment (jpaddison3) [inline packages/lesswrong/lib/collections/users/helpers.ts:353]: I'm surprised at this addition.

File: packages/lesswrong/lib/collections/users/helpers.ts [hunk 3]
```diff
@@ -306,47 +307,84 @@ export const userGetLocation = (currentUser: UsersCurrent|DbUser|null): {
...
   const placeholderLat = 37.871853
   const placeholderLng = -122.258423
+  const defaultLocation = {lat: placeholderLat, lng: placeholderLng, loading: false, known: false, label: null}
   
   const currentUserLat = currentUser && currentUser.mongoLocation && currentUser.mongoLocation.coordinates[1]
   const currentUserLng = currentUser && currentUser.mongoLocation && currentUser.mongoLocation.coordinates[0]
 
   const [location, setLocation] = useState(() => {
-    if (currentUserLat && currentUserLng) {
+    if (currentUserLat && currentUserLng && currentUser?.location) {
```
Comment (jpaddison3) [inline packages/lesswrong/lib/collections/users/helpers.ts:351]: Do we need this &&? What if the user just has a lat-lon set?

File: packages/lesswrong/lib/collections/users/helpers.ts [hunk 4]
```diff
@@ -306,47 +307,84 @@ export const userGetLocation = (currentUser: UsersCurrent|DbUser|null): {
...
+} => {
+  // default is Berkeley, CA
   const placeholderLat = 37.871853
   const placeholderLng = -122.258423
+  const defaultLocation = {lat: placeholderLat, lng: placeholderLng, loading: false, known: false, label: null}
   
   const currentUserLat = currentUser && currentUser.mongoLocation && currentUser.mongoLocation.coordinates[1]
   const currentUserLng = currentUser && currentUser.mongoLocation && currentUser.mongoLocation.coordinates[0]
 
   const [location, setLocation] = useState(() => {
```
Comment (jpaddison3) [inline packages/lesswrong/lib/collections/users/helpers.ts:350]: Note: The location var here is an object but there's also user.location which is a string. This overloading confused me for a bit. Not sure if worth doing anything about.

**Review by jpaddison3 — APPROVED** [review]
ASTLAC

File: packages/lesswrong/lib/collections/users/helpers.ts (hunk 2 above)
Comment (s-cheng) [reply packages/lesswrong/lib/collections/users/helpers.ts:353]: Can you expand on this? In what way are you surprised, and what are you surprised about?

File: packages/lesswrong/lib/collections/users/helpers.ts (hunk 3 above)
Comment (s-cheng) [reply packages/lesswrong/lib/collections/users/helpers.ts:351]: Yeah this was a bit of a toss up. I guess it's possible to have the lat/lng and not the location name, and maybe it's better to have the page load faster in that case even if we list an empty string as their location.

File: packages/lesswrong/lib/collections/users/helpers.ts (hunk 4 above)
Comment (s-cheng) [reply packages/lesswrong/lib/collections/users/helpers.ts]: I renamed the state var to `locationData`

File: packages/lesswrong/lib/collections/users/helpers.ts (hunk 2 above)
Comment (jpaddison3) [reply packages/lesswrong/lib/collections/users/helpers.ts]: I'm not sure what I was saying. No points for clarity here. I think maybe it was the location object vs location string thing.

File: packages/lesswrong/lib/collections/users/helpers.ts (hunk 3 above)
Comment (jpaddison3) [reply packages/lesswrong/lib/collections/users/helpers.ts]: Isn't that quite common? For all users who visited other pages before your code?

Comment (s-cheng) [reply packages/lesswrong/lib/collections/users/helpers.ts]: I think the Community page got on the order of 10s of users between me adding location saving and now, so I wouldn't say it's common :)

**Pull request merged**